File: .leetcode/15.三数之和.c
/*
 * @lc app=leetcode.cn id=15 lang=c
 *
 * [15] 三数之和
 */

// @lc code=start
/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
int** threeSum(int* nums, int numsSize, int* returnSize/*一共这么多个不同的三元组*/, int** returnColumnSizes) {
    int **three;
    *returnSize=0;
    three=malloc(sizeof(int *)*numsSize);
    for(int i=0;i<numsSize;i++){
        three[i]=malloc(sizeof(int)*3);
    }

    for(int i=0;i<numsSize-2;i++){
        for(int j=i+1;j<numsSize-1;j++){
            for(int k=j+1;k<numsSize;k++){
                if(nums[i]+nums[j]+nums[k]==0){
                    three[*returnSize][0]=nums[i];
                    three[*returnSize][1]=nums[j];
                    three[*returnSize][2]=nums[k];
                    (*returnSize)++;               
            }
        }
    }
    return three;
}
// @lc code=end

File: .leetcode/15.三数之和.py
#
# @lc app=leetcode.cn id=15 lang=python
#
# [15] 三数之和
#

# @lc code=start
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        a=[]
        nums=sorted(nums)
        for i in range(len(nums)-2):
            if i>0 and nums[i]==nums[i-1]:
                continue
            else:
                j=i+1
                k=len(nums)-1
                target=-nums[i] 
                while j<k:   
                    if nums[j] + nums[k] == target:
                        a.append([nums[i], nums[j], nums[k]])
                        # 跳过重复的 nums[j] 和 nums[k]
                        while j < k and nums[j] == nums[j+1]:
                            j += 1
                        while j < k and nums[k] == nums[k-1]:
                            k -= 1
                        j += 1
                        k -= 1
                    elif nums[j]+nums[k]>target:
                        k-=1
                    else:
                        j+=1
        return a
# @lc code=end

